File: src/structural/facade/python/classes/builder_facade_decorator.py
from classes.facade.builder_facade import BuilderFacade


class BuilderFacadeDecorator():
    def __init__(self):
        self.facade = BuilderFacade()

    def before(self):
        self.facade.main_dish_builder.reset()
        self.facade.vegan_dish_builder.reset()

    def make_meal_with_two_builder(self):
        self.before()
        self.facade.make_meal_with_two_builder()
        print('\n >> Finish make_meal_with_two_builder\n')

    def make_meal_without_builder(self):
        self.before()
        self.facade.make_meal_without_builder()
        print('\n >> Finish make_meal_without_builder\n')

File: src/structural/facade/typescript/classes/BuilderFacadeDecorator.ts
import { BuilderFacade } from './facade/BuilderFacade';
import { BuilderFacadeProtocol } from './interfaces/BuilderFacadeProtocol';

export class BuilderFacadeDecorator implements BuilderFacadeProtocol {
  private facade = new BuilderFacade();

  private before(): void {
    this.facade.mainDishBuilder.reset();
  }

  makeMealWithTwoBuilders(): void {
    this.before();
    this.facade.makeMealWithTwoBuilders();
    console.log('\n >> Finish makeMealWithTwoBuilders\n');
  }
  makeMealWithoutBuilder(): void {
    this.before();
    this.facade.makeMealWithoutBuilder();
    console.log('\n >> Finish makeMealWithoutBuilder\n');
  }
}
